src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -258,10 +292,28 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
 
         public async Task<UpsertOutcome> UpsertAsync(ResourceWrapperOperation resource, CancellationToken cancellationToken)
         {
-            // TODO: Remove if when Merge is min supported version
-            if (_schemaInformation.Current >= SchemaVersionConstants.Merge && _mergeResourcesFeatureFlag.IsEnabled())
+            bool isBundleOperation = _bundleOrchestrator.IsEnabled && resource.BundleOperationId != null;
+
+            // TODO: Remove it when Merge is min supported version
+            if (_schemaInformation.Current >= SchemaVersionConstants.Merge && _mergeResourcesFeatureFlag.IsEnabled() && !isBundleOperation)
             {
-                return (await MergeAsync(new List<ResourceWrapperOperation> { resource }, cancellationToken)).First().Value;
+                var mergeOutcome = await MergeAsync(new List<ResourceWrapperOperation> { resource }, cancellationToken);
+                DataStoreOperationOutcome dataStoreOperationOutcome = mergeOutcome.First().Value;
+
+                if (dataStoreOperationOutcome.IsOperationSuccessful)
+                {
+                    return dataStoreOperationOutcome.UpsertOutcome;
+                }
+                else
+                {
+                    throw dataStoreOperationOutcome.Exception;
+                }
+            }
+            else if (_schemaInformation.Current >= SchemaVersionConstants.Merge && _mergeResourcesFeatureFlag.IsEnabled() && isBundleOperation)
+            {
+                IBundleOrchestratorOperation bundleOperation = _bundleOrchestrator.GetOperation(resource.BundleOperationId.Value);
```

Comment:
NIT: Isn't readable to create an `if (isBundleOperation) {} else {}`, instead of adding a new value in the same line?